File: ml/suara_ml/textnorm.py
"""Normalisasi dan tokenisasi teks ulasan.

Semua aturan di sini sengaja dibuat sederhana dan deterministik supaya bisa
ditiru persis di TypeScript (web/lib/textnorm.ts). Kalau aturan diubah di sini,
ubah juga di sana lalu jalankan uji paritas.
"""

from __future__ import annotations

import json
import re
import unicodedata
from functools import lru_cache
from pathlib import Path

SLANG_PATH = Path(__file__).with_name("slang_id.json")

URL_RE = re.compile(r"(https?://\S+|www\.\S+)")
REPEAT_RE = re.compile(r"(.)\1{2,}")
TOKEN_RE = re.compile(r"[a-z0-9]+|[☀-➿\U0001f300-\U0001faff]")


@lru_cache(maxsize=1)
def slang_map() -> dict[str, str]:
    with open(SLANG_PATH, encoding="utf-8") as f:
        return json.load(f)


def normalize(text: str) -> str:
    """Teks mentah ke teks bersih (huruf kecil, tanpa URL, huruf berulang dipangkas)."""
    if not isinstance(text, str):
        return ""
    t = unicodedata.normalize("NFKC", text).lower()
    t = URL_RE.sub(" ", t)
    t = REPEAT_RE.sub(r"\1\1", t)
    return t


def tokens(text: str, use_slang: bool = True) -> list[str]:
    """Token kata setelah normalisasi. Slang dipetakan ke bentuk baku."""
    toks = TOKEN_RE.findall(normalize(text))
    if use_slang:
        sm = slang_map()
        out: list[str] = []
        for tok in toks:
            out.extend(sm.get(tok, tok).split())
        return out
    return toks


def word_ngrams(toks: list[str], n_max: int = 2) -> list[str]:
    feats = list(toks)
    for n in range(2, n_max + 1):
        feats.extend(" ".join(toks[i : i + n]) for i in range(len(toks) - n + 1))
    return feats


def char_ngrams(toks: list[str], n_min: int = 2, n_max: int = 5) -> list[str]:
    """N-gram karakter per token, token dibungkus spasi (mirip char_wb sklearn)."""
    feats: list[str] = []
    for tok in toks:
        w = " " + tok + " "
        L = len(w)
        for n in range(n_min, n_max + 1):
            if n > L:
                break
            feats.extend(w[i : i + n] for i in range(L - n + 1))
    return feats


_DIGITS_RE = re.compile(r"\d{6,}")
_EMAIL_RE = re.compile(r"[\w.+-]+@[\w-]+\.[\w.]+")


def mask_personal(text: str) -> str:
    """Samarkan deret angka panjang dan email sebelum teks ulasan dipublikasikan."""
    return _DIGITS_RE.sub("[angka]", _EMAIL_RE.sub("[email]", text))

File: web/lib/textnorm.ts
// Port persis dari ml/suara_ml/textnorm.py. Kalau salah satu diubah, ubah keduanya
// lalu jalankan uji paritas (npm run parity).

import slang from "./slang_id.json";

const SLANG = slang as Record<string, string>;
const URL_RE = /(https?:\/\/\S+|www\.\S+)/gu;
const REPEAT_RE = /(.)\1{2,}/gu;
const TOKEN_RE = /[a-z0-9]+|[☀-➿\u{1F300}-\u{1FAFF}]/gu;

export function normalize(text: string): string {
  if (typeof text !== "string") return "";
  let t = text.normalize("NFKC").toLowerCase();
  t = t.replace(URL_RE, " ");
  t = t.replace(REPEAT_RE, "$1$1");
  return t;
}

export function tokens(text: string, useSlang = true): string[] {
  const raw = normalize(text).match(TOKEN_RE) ?? [];
  if (!useSlang) return raw;
  const out: string[] = [];
  for (const tok of raw) {
    const mapped = Object.prototype.hasOwnProperty.call(SLANG, tok) ? SLANG[tok] : tok;
    for (const part of mapped.split(/\s+/)) if (part) out.push(part);
  }
  return out;
}

export function wordNgrams(toks: string[], nMax = 2): string[] {
  const feats = [...toks];
  for (let n = 2; n <= nMax; n++) {
    for (let i = 0; i + n <= toks.length; i++) feats.push(toks.slice(i, i + n).join(" "));
  }
  return feats;
}

/** N-gram karakter per token (dihitung per code point, sama seperti Python). */
export function charNgramsOf(tok: string, nMin = 2, nMax = 5): string[] {
  const w = Array.from(" " + tok + " ");
  const L = w.length;
  const feats: string[] = [];
  for (let n = nMin; n <= nMax; n++) {
    if (n > L) break;
    for (let i = 0; i + n <= L; i++) feats.push(w.slice(i, i + n).join(""));
  }
  return feats;
}
